smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,457 @@
+                                                          instances=available_insts,
+                                                          run_history=run_history,
+                                                          budget=budget,
+                                                          inc_sum_cost=inc_sum_cost,
+                                                          first_run=first_run)
+
+                # if all challengers were capped, then stop intensification
+                if not curr_challengers:
+                    self.logger.info("All configurations have been eliminated!"
+                                     "Interrupting optimization run and returning current incumbent")
+                    inc_perf = run_history.get_cost(incumbent)
+                    return incumbent, inc_perf
+
+            except BudgetExhaustedException:
+                # Returning the final incumbent selected so far because we ran out of optimization budget
+                self.logger.debug("Budget exhausted; "
+                                  "Interrupting optimization run and returning current incumbent")
+                inc_perf = run_history.get_cost(incumbent)
+                return incumbent, inc_perf
+
+            # selecting the top 'k' challengers for the next iteration
+            if next_n_chal > 0:
+                curr_challengers = self._top_k(curr_challengers, run_history, k=next_n_chal)
+            else:
+                # returning at least 1 configuration for next iteration since the incumbent has to run on all instances
+                self.logger.info('Ran out of configurations! Returning the best from what was seen so far.')
+                curr_challengers = self._top_k(curr_challengers, run_history, k=1)
+
+            first_run = False
+
+            # increase budget for next iteration
+            prev_budget = budget
+            budget = budget * self.eta
+
+        # select best challenger from the SH run
+        best_challenger = curr_challengers[0]
+        self.logger.debug("Best challenger from successive halving run - %s" % (str(best_challenger)))
+
+        # compare challenger with current incumbent
+        if incumbent is None:  # first intensify run from initial design
+            new_incumbent = best_challenger
+            inc_perf = run_history.get_cost(best_challenger)
+            self.logger.info("First Incumbent found! Cost of incumbent is (%.4f)" % inc_perf)
+            self.logger.info("incumbent configuration: %s" % str(best_challenger))
+            if log_traj:
+                # adding incumbent entry
+                self.stats.inc_changed += 1  # first incumbent
+                self.traj_logger.add_entry(train_perf=inc_perf,
+                                           incumbent_id=self.stats.inc_changed,
+                                           incumbent=new_incumbent)
+
+        else:
+            new_incumbent = self._compare_configs(incumbent, best_challenger,
```

Comment:
I'm wondering how we would define incumbent best. I think only returning an incumbent after the successive halving procedure is completed is too conservative. Therefore, the question is whether the incumbent is defined as a configuration+budget pair or only as a configuration? Only in the latter case we want to only compare in the end, otherwise we want to compare it in between.

Reply:
This will be part of a later pull request